### PR TITLE
null is not an accepted type so convert it to an accepted type

### LINKED ISSUE #3520: NPEs when initializing items
The demo setup of openHAB now results in such exceptions in the log:

```
2017-05-25 17:28:58.990 [ERROR] [ore.internal.events.OSGiEventManager] - Dispatching/filtering event for subscriber 'org.eclipse.smarthome.core.events.EventSubscriber' failed: java.lang.IllegalArgumentException: The argument 'state' must not be null or empty.
java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: The argument 'state' must not be null or empty.
        at java.util.concurrent.FutureTask.report(FutureTask.java:122)[:1.8.0_112]
        at java.util.concurrent.FutureTask.get(FutureTask.java:206)[:1.8.0_112]
        at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:194)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:83)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:67)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager.dispatchESHEvent(OSGiEventManager.java:188)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager.handleEvent(OSGiEventManager.java:163)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager.handleEvent(OSGiEventManager.java:146)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.apache.felix.eventadmin.impl.handler.EventHandlerProxy.sendEvent(EventHandlerProxy.java:415)[6:org.apache.karaf.services.eventadmin:4.0.8]
        at org.apache.felix.eventadmin.impl.tasks.HandlerTask.run(HandlerTask.java:90)[6:org.apache.karaf.services.eventadmin:4.0.8]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_112]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_112]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_112]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_112]
        at java.lang.Thread.run(Thread.java:745)[:1.8.0_112]
Caused by: java.lang.IllegalArgumentException: The argument 'state' must not be null or empty.
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:122)[12:com.google.guava:18.0.0]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.assertValidArguments(ItemEventFactory.java:370)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.items.events.ItemEventFactory.createStateChangedEvent(ItemEventFactory.java:272)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.items.GenericItem.sendStateChangedEvent(GenericItem.java:215)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.items.GenericItem.applyState(GenericItem.java:209)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.library.items.DimmerItem.setState(DimmerItem.java:74)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.items.ItemUpdater.receiveUpdate(ItemUpdater.java:70)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.items.events.AbstractItemEventSubscriber.receive(AbstractItemEventSubscriber.java:44)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager$1.call(OSGiEventManager.java:192)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager$1.call(OSGiEventManager.java:1)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:181)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        ... 4 more
2017-05-25 17:28:59.064 [ERROR] [ore.internal.events.OSGiEventManager] - Dispatching/filtering event for subscriber 'org.eclipse.smarthome.core.events.EventSubscriber' failed: java.lang.NullPointerException
java.util.concurrent.ExecutionException: java.lang.NullPointerException
        at java.util.concurrent.FutureTask.report(FutureTask.java:122)[:1.8.0_112]
        at java.util.concurrent.FutureTask.get(FutureTask.java:206)[:1.8.0_112]
        at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:194)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:83)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:67)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager.dispatchESHEvent(OSGiEventManager.java:188)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager.handleEvent(OSGiEventManager.java:163)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager.handleEvent(OSGiEventManager.java:146)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.apache.felix.eventadmin.impl.handler.EventHandlerProxy.sendEvent(EventHandlerProxy.java:415)[6:org.apache.karaf.services.eventadmin:4.0.8]
        at org.apache.felix.eventadmin.impl.tasks.HandlerTask.run(HandlerTask.java:90)[6:org.apache.karaf.services.eventadmin:4.0.8]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_112]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_112]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_112]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_112]
        at java.lang.Thread.run(Thread.java:745)[:1.8.0_112]
Caused by: java.lang.NullPointerException
        at org.eclipse.smarthome.core.items.GenericItem.applyState(GenericItem.java:208)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.library.items.DimmerItem.setState(DimmerItem.java:74)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.items.ItemUpdater.receiveUpdate(ItemUpdater.java:70)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.items.events.AbstractItemEventSubscriber.receive(AbstractItemEventSubscriber.java:44)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager$1.call(OSGiEventManager.java:192)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager$1.call(OSGiEventManager.java:1)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:181)[99:org.eclipse.smarthome.core:0.9.0.201705251309]
        ... 4 more
```

See also https://community.openhab.org/t/java-lang-nullpointerexception-every-few-seconds/28944

### PR DESCRIPTION
**Pull request by martinvw**
null is not an accepted type so convert it to an accepted type

Overcomes NPE's in a places where the state is broadcasted to

Fixes #3520

### CODE REVIEW

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java [hunk 1]
```diff
@@ -75,14 +75,18 @@ public static void assertValidItemName(String itemName) throws IllegalArgumentEx
...
+        }
+
         if (item != null && !isAccepted(item, state)) {
             for (Class<? extends State> acceptedType : item.getAcceptedDataTypes()) {
                 State convertedState = state.as(acceptedType);
                 if (convertedState != null) {
                     LoggerFactory.getLogger(ItemUtil.class).debug("Converting {} '{}' to {} '{}' for item '{}'",
-                            state.getClass().getSimpleName(), state.toString(),
-                            convertedState.getClass().getSimpleName(), convertedState.toString(), item.getName());
+                            state.getClass().getSimpleName(), state, convertedState.getClass().getSimpleName(),
```
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: You should not call toString on logging arguments, just pass them on as an object they will be converted by logging framework if the log level applies

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java [hunk 2]
```diff
@@ -75,14 +75,18 @@ public static void assertValidItemName(String itemName) throws IllegalArgumentEx
...
+    public static State convertToAcceptedState(State state, Item item) {
+        if (state == null) {
+            return UnDefType.NULL;
+        }
+
         if (item != null && !isAccepted(item, state)) {
             for (Class<? extends State> acceptedType : item.getAcceptedDataTypes()) {
                 State convertedState = state.as(acceptedType);
                 if (convertedState != null) {
                     LoggerFactory.getLogger(ItemUtil.class).debug("Converting {} '{}' to {} '{}' for item '{}'",
```
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: imho this is a real strange solution to overcome a static logger. Now you are going to re-fetch the logger from the factory each time you pass by this debug (!!!) statement which is really often for this specific line. The coding guidelines leave room to have static loggers if required.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java [hunk 3]
```diff
@@ -75,14 +75,18 @@ public static void assertValidItemName(String itemName) throws IllegalArgumentEx
         }
     }
 
-    public static State convertToAcceptedState(final State state, final Item item) {
+    public static State convertToAcceptedState(State state, Item item) {
+        if (state == null) {
```
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: This is the core of the actual change :-)

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java [hunk 4]
```diff
@@ -44,11 +47,8 @@ private ItemUtil() {
      * @return true if the specified name is a valid item name, otherwise false
      */
     public static boolean isValidItemName(final String itemName) {
-        if (itemName != null && !itemName.isEmpty()) {
-            return itemName.matches("[a-zA-Z0-9_]*");
-        }
+        return StringUtils.isNotEmpty(itemName) && itemName.matches("[a-zA-Z0-9_]*");
```
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: Sorry could not leave it like that :-)

Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: Btw, do we think that itemName of 0 characters is valid....?

**Comment by martinvw** [comment]
@kaikreuzer imho this should solve the problem which was mentioned in the issue #3520 how were you able to reproduce it?

And on the community forum: https://community.openhab.org/t/java-lang-nullpointerexception-every-few-seconds/28944

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java (hunk 4 above)
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: 👍  and no, AFAIK it needs to be a least one character long

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java (hunk 3 above)
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: I wonder how `null` actually gets in here. IMHO we should fix the root cause of it first. Still, it's valid to make it fail-safe here.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java (hunk 1 above)
Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: 👍

**Comment by sjka** [comment]
The main question remains: What is the root-cause? Items should initially have a `UnDefType.NULL` state, so something recently must have changed which breaks with this assumption. Any clues?

**Comment by sjka** [comment]
GroupItems without a function maybe? Then it could have been https://github.com/eclipse/smarthome/pull/3411

**Comment by kaikreuzer** [comment]
I think this is a likely a correct assumption...

**Comment by martinvw** [comment]
If I follow up the call-stack, I arrive here:
ItemUtil.convertToAcceptedState
called from 
ThingManager line 128 thingHandlerCallback.stateUpdated
called from
BaseThingHandler.updateState

Why can't it be a binding which just call's:

`updateState(channelUID, null);`

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java (hunk 4 above)
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: > and no, AFAIK it needs to be a least one character long

@kaikreuzer agree? Should it be a separate PR?

**Comment by sjka** [comment]
> Why can't it be a binding which just call's:
updateState(channelUID, null);

Because then (as bad as it is...) it would have failed with an NPE in 
[ItemUtil line 94](https://github.com/eclipse/smarthome/blob/69393152219b85ae0e9efffda9793271345b7625/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java#L94), if I'm not mistaken.

**Comment by martinvw** [comment]
Well spotted, yes I agree on that, so my change most likely doesn't fix the issue

**Comment by martinvw** [comment]
I have my sister visiting, but I'll add changes later to fix setting of the state of e group.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java (hunk 4 above)
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: Certainly agree. Yes, to be clean, you can do it in a separate PR.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java [hunk 5]
```diff
@@ -44,11 +47,8 @@ private ItemUtil() {
...
```
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: hey, you missed to remove this empty line :-)

**Comment by martinvw** [comment]
I don't see it coming from the mentioned PR either (no call to setState which might contain `null`), what do you think about this one:
[GroupItem.java#L329](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/GroupItem.java#L329)

**Comment by martinvw** [comment]
> I don't see it coming from the mentioned PR either (no call to setState which might contain null), what do you think about this one: GroupItem.java#L329

Then I'm most likely missing something because my branch contains no function which can return null either. Can you point out the line where you think its actually assigning null to the `oldState`?

**Comment by martinvw** [comment]
Another likely suspect, #3438

https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/types/State.java#L29

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java (hunk 2 above)
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: @kaikreuzer what is your view on this specific solution

Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: static logger would definitely be ok for me here

**Comment by martinvw** [comment]
@triller-telekom can you take a look at NPE in combination with the change and line to which I referred, do you think this might be the cause of this? Thanks!

**Comment by kaikreuzer** [comment]
You are right, the issue is definitely in https://github.com/eclipse/smarthome/commit/69393152219b85ae0e9efffda9793271345b7625#diff-543ecd419447346e6a177db8638e3ad0R74 - the `as()` method is allowed to return null and if you feed in an UNDEF at that place, the state will (now) be set to null internally. This worked in the past as UNDEF wasn't a Convertable state and thus was set directly.

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java (hunk 3 above)
Comment (kaikreuzer) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: @martinvw I think we should not silently swallow this situation, but log an error (incl. stacktrace) nonetheless to highlight that there is some bug somewhere in the code as we should never get here.

**Comment by martinvw** [comment]
So do you mean an UNDEF is converted to null, should we change the default implementation in that case?

Or should UnDefType override that method with `return this;`

File: bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java (hunk 4 above)
Comment (martinvw) [inline bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/items/ItemUtil.java]: I just see that we just did a isNotEmpty check so only the regex looked misleading but nothing will change :-D

**Comment by kaikreuzer** [comment]
Not sure how we best get back the behavior we had with the "Convertibles" - I'd leave it to @sjka & @triller-telekom to decide as they both worked on that.
